### DISCUSSION TITLE
Error running cache:clear and seeder with cached queries

### DISCUSSION THREAD

**Discussion by softnex** [spatie/laravel-multitenancy#328] [Q&A]
I have multitenancy with multi-database.

When running the cache:clear command (both for a specific tenant or for all) I get the following error.

The same happens if I run one seeder with php artisan and it has cached queries.

Example:

php artisan tenants:artisan "cache:clear" --tenant=1

Running command for tenant `client one` (id: 1)...
---------------------------------------------------------
[2022-02-09T18:02:21.575521-03:00] local.ERROR: Call to a member function prepare() on null {"exception":"[object] (Error(code: 0): Call to a member function prepare() on null at /project/vendor/laravel/framework/src/Illuminate/Database/Connection.php:522)"} []

   Error 

  Call to a member function prepare() on null

  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:522
    518▕ 
    519▕             // For update or delete statements, we want to get the number of rows affected
    520▕             // by the statement and return that back to the developer. We'll first need
    521▕             // to execute the statement and then we'll use PDO to fetch the affected.
  ➜ 522▕             $statement = $this->getPdo()->prepare($query);
    523▕ 
    524▕             $this->bindValues($statement, $this->prepareBindings($bindings));
    525▕ 
    526▕             $statement->execute();

      +43 vendor frames 
  44  artisan:37
      Illuminate\Foundation\Console\Kernel::handle()

**Comment by softnex** [comment]
Forgot to comment. The cache is in database

**Reply by masterix21** [reply]
@softnex, please post your full stacktrace. Thanks

**Comment by softnex** [comment]
As I'm running from the command line, the only message is the one I posted.

When I use the command to generate seeder, which uses cached queries, the error also occurs

Seeder example.
  php artisan tenants:artisan "db:seed --class=RegistrySeeder" --tenant=1


Returns the error below:

-------------------------------
Running command for tenant `client one` (id: 1)...
---------------------------------------------------------
[2022-02-10T08:02:32.690464-03:00] local.ERROR: Call to a member function prepare() on null {"exception":"[object] (Error(code: 0): Call to a member function prepare() on null at /project/vendor/laravel/framework/src/Illuminate/Database/Connection.php:368)"} []

   Error 

  Call to a member function prepare() on null

  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:368
    364▕             // For select statements, we'll simply execute the query and return an array
    365▕             // of the database result set. Each element in the array will be a single
    366▕             // row from the database table, and will either be an array or objects.
    367▕             $statement = $this->prepared(
  ➜ 368▕                 $this->getPdoForSelect($useReadPdo)->prepare($query)
    369▕             );
    370▕ 
    371▕             $this->bindValues($statement, $this->prepareBindings($bindings));
    372▕ 

      +13 vendor frames 
  14  app/Repositories/BaseRepository.php:201
      Illuminate\Support\Facades\Facade::__callStatic()

  15  app/Services/BaseService.php:48
      App\Repositories\BaseRepository::get()


BaseRepository.php line 201 contains the following:

       return Cache::remember('query_name', Carbon::now()->addMinutes(60), function() use ($filters, $fields, $order) {
            return $this->getBuilder($filters, $fields, $order)->get();
        });
        


Apparently it is not setting the default connection to the database when it is run by the command line, but if I put the command dd(app('currentTenant')); at the beginning of the seeder I get the Model correctly as shown below.
      
        
Running command for tenant `client one` (id: 1)...
---------------------------------------------------------
Seeding: Database\Seeders\Messages\RegistrySeeder
Spatie\Multitenancy\Models\Tenant^ {#2843
  #connection: "landlord"
  #table: "tenants"
  #primaryKey: "id"
  #keyType: "int"
  +incrementing: true
  #with: []
  #withCount: []
  +preventsLazyLoading: false
  #perPage: 15
  +exists: true
  +wasRecentlyCreated: false
  #escapeWhenCastingToString: false
  #attributes: array:6 [
    "id" => 1
    "name" => "Client One"
    "domain" => "/php80/clientone.project"
    "database" => "project_clientone"
    "created_at" => "2021-09-03 16:02:17"
    "updated_at" => "2021-09-03 16:02:17"
  ]
  #original: array:6 [
    "id" => 1
    "name" => "Client One"
    "domain" => "/php80/clientone.project"
    "database" => "project_clientone"
    "created_at" => "2021-09-03 16:02:17"
    "updated_at" => "2021-09-03 16:02:17"
  ]
  #changes: []
  #casts: []
  #classCastCache: []
  #attributeCastCache: []
  #dates: []
  #dateFormat: null
  #appends: []
  #dispatchesEvents: []
  #observables: []
  #relations: []
  #touches: []
  +timestamps: true
  #hidden: []
  #visible: []
  #fillable: []
  #guarded: array:1 [
    0 => "*"
  ]
}



The error definitely happens on line 368 of the source vendor/laravel/framework/src/Illuminate/Database/Connection.php
with the following query being executed.

select * from `cache` where `key` = ? limit 1

The command $this->getDatabaseName() return null, that is, it seems to have not set the connection, only for the cache query.

My cache.php configuration looks like this

         'database' => [
             'driver' => 'database',
             'table' => 'cache',
             'connection' => env('DB_CONNECTION'),
             'lock_connection' => null,
         ],

And my .env is as determined by the documentation, with DB_DATABASE null because it is mounted in multi-tenancy.

DB_CONNECTION=tenant
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=
DB_USERNAME=xxxx
DB_PASSWORD=xxxxx
